localstack/services/s3/v3/provider.py

```diff
@@ -0,0 +1,1231 @@
+            delete_marker = S3DeleteMarker(key=key, version_id=delete_marker_id)
+            # TODO: verify with Suspended bucket? does it override last version or still append?? big question
+            #  I think it puts a delete marker with a `null` VersionId, which deletes the object under
+            s3_bucket.objects.set(key, delete_marker)
+            return DeleteObjectOutput(VersionId=delete_marker.version_id, DeleteMarker=True)
+
+        if key not in s3_bucket.objects:
+            return DeleteObjectOutput()
+
+        if not (found_object := s3_bucket.objects.pop(object_key=key, version_id=version_id)):
+            raise InvalidArgument(
+                "Invalid version id specified",
+                ArgumentName="versionId",
+                ArgumentValue=version_id,
+            )
+
+        response = DeleteObjectOutput(VersionId=found_object.version_id)
+
+        if isinstance(found_object, S3DeleteMarker):
+            response["DeleteMarker"] = True
+        else:
+            self._storage_backend.remove(bucket, found_object)
+
+        return response
+
+    def delete_objects(
+        self,
+        context: RequestContext,
+        bucket: BucketName,
+        delete: Delete,
+        mfa: MFA = None,
+        request_payer: RequestPayer = None,
+        bypass_governance_retention: BypassGovernanceRetention = None,
+        expected_bucket_owner: AccountId = None,
+        checksum_algorithm: ChecksumAlgorithm = None,
+    ) -> DeleteObjectsOutput:
+        store = self.get_store(context.account_id, context.region)
+        if not (s3_bucket := store.buckets.get(bucket)):
+            raise NoSuchBucket("The specified bucket does not exist", BucketName=bucket)
+
+        objects: list[ObjectIdentifier] = delete.get("Objects")
+        if not objects:
+            raise MalformedXML()
+
+        # TODO: max 1000 delete at once? test against AWS?
+        # TODO: implement ByPassGovernance
+        # TODO: implement Locking error
+
+        quiet = delete.get("Quiet", False)
+        deleted = []
+        errors = []
+
+        for to_delete_object in objects:
+            # TODO: beware of key encoding (XML?)
+            object_key = to_delete_object.get("Key")
+            version_id = to_delete_object.get("VersionId")
+            if s3_bucket.versioning_status is None:
+                if version_id and version_id != "null":
+                    errors.append(
+                        Error(
+                            Code="NoSuchVersion",
+                            Key=object_key,
+                            Message="The specified version does not exist.",
+                            VersionId=version_id,
+                        )
+                    )
+                    continue
+
+                found_object = s3_bucket.objects.pop(object_key, None)
+                if found_object:
+                    self._storage_backend.remove(bucket, found_object)
```

Comment:
would it make sense to allow to pass a tuple/list of objects to remove? or add a method `remove_many` or something? that way the underlying storage system has an opportunity to optimize a batched remove.

Reply:
No yes it would make sense. Will add 👍